File: Extensions/Tessa.Extensions.Default.Client/Files/RestoreFileControlExtension.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrPermissions;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform.Runtime;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;
using Tessa.UI.Cards.Forms;
using Tessa.UI.Cards.Tasks;
using Tessa.UI.Files;
using Tessa.UI.Menu;
using Tessa.Views;
using Tessa.Views.Metadata;
using Tessa.Views.Metadata.Criteria;

namespace Tessa.Extensions.Default.Client.Files
{
    /// <summary>
    /// Расширение, добавляющее в меню файлового контрола пункт для восстановления файла из корзины.
    /// </summary>
    public sealed class RestoreFileControlExtension :
        FileControlExtension
    {
        #region Fields

        private readonly IUIHost uiHost;
        private readonly ISession session;
        private readonly IViewService viewService;
        private readonly ICardFileBackupSettings cardFileBackupSettings;

        #endregion

        #region Constructors

        /// <summary>
        /// Создает экземпляр класса <see cref="RestoreFileControlExtension"/>.
        /// </summary>
        /// <param name="uiHost"><inheritdoc cref="IUIHost" path="/summary"/></param>
        /// <param name="session"><inheritdoc cref="ISession" path="/summary"/></param>
        /// <param name="viewService"><inheritdoc cref="IViewService" path="/summary"/></param>
        /// <param name="cardFileBackupSettings"><inheritdoc cref="ICardFileBackupSettings" path="/summary"/></param>
        public RestoreFileControlExtension(
            IUIHost uiHost,
            ISession session,
            IViewService viewService,
            ICardFileBackupSettings cardFileBackupSettings)
        {
            this.uiHost = NotNullOrThrow(uiHost);
            this.session = NotNullOrThrow(session);
            this.viewService = NotNullOrThrow(viewService);
            this.cardFileBackupSettings = NotNullOrThrow(cardFileBackupSettings);
        }

        #endregion

        #region Base Overrides

        /// <inheritdoc/>
        public override async Task OpeningMenu(IFileControlExtensionContext context)
        {
            var cardModel = UIContext.Current.CardEditor?.CardModel;
            if (cardModel is null
                || cardModel.InSpecialMode()
                || IsCardTaskFileControl(cardModel, context.Control)
                || !await this.cardFileBackupSettings.CanDeleteWithBackupAsync(
                    context.ValidationResult,
                    context.CancellationToken)
                || context.Actions.FirstOrDefault(a => a.Name is FileMenuActionNames.Upload)
                    is not { IsCollapsed: false } uploadAction)
            {
                return;
            }

            var uploadActionIndex = context.Actions.IndexOf(uploadAction);

            context.Actions.Insert(
                uploadActionIndex + 1,
                new MenuAction(
                    FileMenuActionNames.Restore,
                    "$UI_Controls_FilesControl_RestoreFile",
                    context.Icons.Get("Thin119"),
                    new DelegateCommand(async _ => await this.RestoreActionAsync(context.CancellationToken)),
                    isCollapsed: !context.Control.Container.Permissions.CanAdd
                ));
        }

        #endregion

        #region Private Methods

        private static bool IsCardTaskFileControl(ICardModel cardModel, IFileControl fileControl) =>
            cardModel.MainForm is DefaultFormTabWithTasksViewModel mainForm
            && mainForm.Tasks.OfType<TaskViewModel>().Any(task =>
                task.TaskModel.ControlBag.OfType<FileListViewModel>().Any(file =>
                    file.FileControl == fileControl));

        private static bool CanRestoreAllDeletedFiles(IUser user, ICardModel cardModel) =>
            user.IsAdministrator()
            || cardModel.Card.TryGetInfo() is { } cardInfo
            && KrToken.TryGet(cardInfo) is { } krToken
            && krToken.HasPermission(KrPermissionFlagDescriptors.RestoreAllDeletedFiles);

        private async Task RestoreActionAsync(CancellationToken cancellationToken)
        {
            var uiContext = UIContext.Current;
            var cardEditor = uiContext.CardEditor;
            var cardModel = cardEditor.CardModel;
            if (cardModel is null)
            {
                return;
            }

            const string viewName = "DeletedFiles";
            List<RequestParameter> viewParameters = new();

            var view = await this.viewService.GetByNameAsync(viewName, cancellationToken)
                ?? throw new InvalidOperationException($"Can not find view with alias \"{viewName}\" at view service.");

            var (viewMetadata, validationResult) = await view.TryGetMetadataAsync(cancellationToken);
            if (viewMetadata is null || !validationResult.IsSuccessful)
            {
                throw new ValidationException(validationResult, new InvalidOperationException($"Can not find metadata for view \"{viewName}\"."));
            }

            if (CanRestoreAllDeletedFiles(this.session.User, cardModel))
            {
                viewParameters.Add(new("ShowAll") { IsTrueCriteriaOperator.Instance });
            }

            viewParameters.Add(new RequestParameter("Card").Add(EqualsToCriteriaOperator.Instance, cardModel.Card.ID));

            var selectedValue = await this.uiHost.ShowViewsDialogAsync(
                viewName,
                viewParameters,
                cancellationToken: cancellationToken);

            var selectedRow = selectedValue?.SelectedRow;
            if (selectedRow is null)
            {
                return;
            }

            var fileName = selectedRow.Get<string>("DeletedName")!;
            var confirmation = await LocalizeFormatAsync("$UI_Controls_FilesControl_RestoreSelectedFileMessage", fileName);
            if (!await TessaDialog.ConfirmAsync(confirmation))
            {
                return;
            }

            var deletedFile = new CardFileDeletedInfo
            {
                RowID = selectedRow.Get<Guid>("DeletedRowID"),
                Deleted = selectedRow.Get<DateTime>("DeletedDate"),
                DeletedByID = selectedRow.Get<Guid>("DeletedByID"),
                DeletedByName = selectedRow.Get<string>("DeletedByName")!
            };

            await cardEditor.SaveCardAsync(
                uiContext,
                request: new CardSavingRequest(
                    CardSavingMode.RefreshOnSuccess,
                    (card, _) =>
                    {
                        card.Info[CardHelper.FilesToRestoreKey] = new List<object> { deletedFile.ToSerializedDictionary() };
                        return ValueTask.CompletedTask;
                    }),
                cancellationToken: cancellationToken);
        }

        #endregion
    }
}

File: WebClient SDK/src/default/files/restoreFileControlExtension.ts
import { extension, inject, ISession, ISession$, IUser, localize } from '@tessa/application';
import {
  CardFileDeletedInfo,
  CardHelper,
  IViewRepository,
  IViewRepository$,
  KrPermissionFlagDescriptors,
  KrToken,
  ViewCriteriaOperators,
  ViewRequestParameter,
  ViewRequestParameterBuilder
} from '@tessa/platform';
import { getTessaIcon } from 'common/utility/uiHelpers';
import { CardSingletonCache, ServerInstanceTypeName } from 'tessa/cards';
import { showConfirm } from 'tessa/ui/tessaDialog/show';
import { CardSavingMode, CardSavingRequest, ICardModel } from 'tessa/ui/cards';
import { FileControlExtension, IFileControl, IFileControlExtensionContext } from 'tessa/ui/files';
import { MenuAction } from 'tessa/ui/menuAction';
import { UIContext } from 'tessa/ui/uiContext';
import { showViewsDialog } from 'tessa/ui/uiHost/showViewsDialog';

/**
 * Расширение, добавляющее в меню файлового контрола пункт для восстановления файла из корзины.
 */
@extension({ name: 'RestoreFileControlExtension' })
export class RestoreFileControlExtension extends FileControlExtension {
  //#region constructors

  constructor(
    @inject(IViewRepository$) private readonly _viewRepository: IViewRepository,
    @inject(ISession$) private readonly _session: ISession
  ) {
    super();
  }

  //#endregion

  //#region base overrides

  public openingMenu(context: IFileControlExtensionContext): void {
    const cardModel = UIContext.current.cardEditor?.cardModel;
    if (
      !cardModel ||
      cardModel?.inSpecialMode ||
      RestoreFileControlExtension.isCardTaskFileControl(context.control)
    ) {
      return;
    }

    const serverInstance = CardSingletonCache.instance.cards.get(ServerInstanceTypeName);
    if (!serverInstance) {
      throw new Error(
        `Can not find card with name "${ServerInstanceTypeName}" at singleton cache.`
      );
    }

    const deleteFilesWithBackup = serverInstance.sections
      .get('ServerInstances')!
      .fields.tryGetBoolean('DeleteFilesWithBackup');

    if (!deleteFilesWithBackup) {
      return;
    }

    const uploadAction = context.actions.find(a => a.name === 'Upload');
    if (!uploadAction || uploadAction.isCollapsed) {
      return;
    }

    const uploadActionIndex = context.actions.indexOf(uploadAction);

    context.actions.splice(
      uploadActionIndex + 1,
      0,
      new MenuAction(
        'Restore',
        '$UI_Controls_FilesControl_RestoreFile',
        getTessaIcon('Thin119'),
        () => this.restoreAction(),
        null,
        !context.control.fileContainer.permissions.canAdd
      )
    );
  }

  //#endregion

  //#region private methods

  private static isCardTaskFileControl(fileControl: IFileControl): boolean {
    return !!fileControl.model.cardTask;
  }

  private static canRestoreAllDeletedFiles(user: IUser, cardModel: ICardModel): boolean {
    if (user.isAdmin) {
      return true;
    }
    const cardInfo = cardModel.card.tryGetInfo();
    return !!(
      cardInfo &&
      KrToken.tryGet(cardInfo)?.hasPermission(KrPermissionFlagDescriptors.RestoreAllDeletedFiles)
    );
  }

  private async restoreAction(): Promise<void> {
    const uiContext = UIContext.current;
    const cardEditor = uiContext.cardEditor;
    const cardModel = cardEditor?.cardModel;
    if (!cardModel) {
      return;
    }

    const viewName = 'DeletedFiles';
    const viewParameters: ViewRequestParameter[] = [];

    const view = await this._viewRepository.getByName(viewName);
    if (!view) {
      throw new Error(`Can not find view with alias "${viewName}" at view service.`);
    }

    const viewMetadata = await view.getMetadata();
    if (!viewMetadata) {
      throw new Error(`Can not find metadata for view "${viewName}".`);
    }

    const cardParameter = viewMetadata.parameters.tryGet('Card');
    if (!cardParameter) {
      throw new Error('Can not find parameter "Card" at view metadata.');
    }

    if (RestoreFileControlExtension.canRestoreAllDeletedFiles(this._session.user, cardModel)) {
      const showAllParameter = viewMetadata.parameters.tryGet('ShowAll');
      if (!showAllParameter) {
        throw new Error('Can not find parameter "ShowAll" at view metadata.');
      }

      viewParameters.push(
        new ViewRequestParameterBuilder()
          .withMetadata(showAllParameter)
          .addCriteria(ViewCriteriaOperators.IsTrue)
          .asRequestParameter()
      );
    }

    viewParameters.push(
      new ViewRequestParameterBuilder()
        .withMetadata(cardParameter)
        .addCriteria(ViewCriteriaOperators.EqualsTo, '', cardModel.card.id)
        .asRequestParameter()
    );

    await showViewsDialog(
      [viewName],
      async ({ selectedRow }) => {
        if (!selectedRow) {
          return;
        }

        const confirmation = localize(
          '$UI_Controls_FilesControl_RestoreSelectedFileMessage',
          selectedRow.get('DeletedName')
        );

        if (!(await showConfirm(confirmation))) {
          return;
        }

        const fileInfo = new CardFileDeletedInfo();
        fileInfo.rowId = selectedRow.get('DeletedRowID');
        fileInfo.deleted = selectedRow.get('DeletedDate');
        fileInfo.deletedById = selectedRow.get('DeletedByID');
        fileInfo.deletedByName = selectedRow.get('DeletedByName');

        await cardEditor.saveCard(
          uiContext,
          undefined,
          new CardSavingRequest(CardSavingMode.RefreshOnSuccess, card => {
            card.info[CardHelper.filesToRestoreKey] = [fileInfo.serializeToStorage()];
          })
        );
      },
      viewParameters
    );
  }

  //#endregion
}
